utils/error_injection.hh

```diff
@@ -93,18 +97,86 @@ extern logging::logger errinj_logger;
  *          }, f);
  *    Expected use case: emulate custom errors like timeouts.
  *
+ * 4. inject_with_handler(name, func)
+ *    Inserts code that can wait for an event.
+ *    Requires func to be a function taking an injection_handler reference and
+ *    returning a future<>.
+ *    Expected use case: wait for an event from tests.
  */
 
 template <bool injection_enabled>
 class error_injection {
     inline static thread_local error_injection _local;
     using handler_fun = std::function<void()>;
 
+    struct received_messages_counter {
```

Comment:
It would be nice to briefly explain the design of how this counting works, listing the main points:
- there is a counter of received messages; it is shared between the `injection_data`, which is created once when enabling an injection on a given shard, and all `injection_handler`s, that are created separately for each firing of this injection
- the counter is incremented when receiving a message from the REST endpoint and the condition variable is signaled
- each `injection_handler` (separate for each firing) stores its own private counter, `_read_messages_counter`
- that private counter is incremented whenever we wait for a message, and compared to the received counter. We sleep on the condition variable if not enough messages were received

Reply:
On the other hand, I just realized that this is actually a reimplementation of semaphore, and not a completely correct one to be fair (e.g. there is a possibility for two concurrent `wait_for_message` to pass through based only on receiving one message)

So maybe we should just use a `seastar::semaphore` (shared with `lw_shared_ptr` like we do with the counter)? Receiving a message `signal()`s, `wait_for_message` `wait()`s.

With a semaphore, we won't need this comment.

Reply:
> e.g. there is a possibility for two concurrent wait_for_message to pass through based only on receiving one message

Isn't it what we agreed to do? That all injection runs should receive a message? You wrote:
> Let me propose an alternative (we discussed it briefly on our 1:1). Each call site of an injection should have a separate injection_handler object created for it. And when you send a message from the test, you deliver it to every injection_handler.

Using `lw_shared_ptr` and `seastar::semaphore` would be a fix and an improvement to my first solution, where one message only notified one handler, but we decided to change that.

Reply:
Sorry I wasn't precise: I meant two concurrent `wait_for_message` happening in **a single firing** of an injection.
I don't know why anyone would want to do anything like this, but you could theoretically do some thing like
```
f1 = wait_for_message();
f2 = wait_for_message();
co_await move(f1);
co_await move(f2);
```

But meh, let's ignore this problem, because I don't think there would be any reason for anyone to do this.

> Using lw_shared_ptr and seastar::semaphore would be a fix and an improvement to my first solution, where one message only notified one handler, but we decided to change that.

Right, I got confused. Indeed using a semaphore would bring back the previous behavior.

So the current solution is OK.